docs/grfid.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" Please adjust this date whenever revising the manpage.
.Dd February 03, 2012
.Dt GRFID 1
.Sh NAME
.Nm GRFID
.Nd A tool to extract the (New)GRF ID from a (New)GRF
.Sh SYNOPSIS
.Op Fl hvm
.Nm grfid Ar file
.Sh DESCRIPTION
GRFID tries to find the (New)GRF identifier of that (New)GRF.
Upon failure an error message is given, otherwise the NewGRF
identifier is printed hexadecimally to stdout.
.Sh OPTIONS
.Bl -tag
.It Fl h
Display the help message.
.It Fl m Ar file
Display the MD5 checksum of the given NewGRF. For GRF container
format 2 this checksum only contains the data section, i.e. it
excludes the sprite section to make it easier to construct NewGRFs
with different zoom levels.
.It Fl v
Display the version.
.El
.Sh SEE ALSO
The documentation in
.Pa /usr/share/doc/grfcodec/
.Sh AUTHOR
GRFID was written by Peter Nelson.
GRFID uses MD5 code that was written by L. Peter Deutsch.
.Pp
This manual page was written by Remko Bijker.
.Sh COPYRIGHT
GRFID is Copyright (C) 2009 by Peter Nelson.
MD5 code is Copyright (C) 1999, 2002 by L. Peter Deutsch.


Changelog
Changes in
- (devel) Detect UPX, fix some race issues in the Makefile, disable
  optimization if GCC 4.5.0 is detected (to work around an optimization bug),
  fix version.h generation if there is a version.h in /usr/include.
- (devel) Fix a number of compiler warnings.
- Clamp the arguments for -w and -h to 65535 as that is the maximum size of a
  PCX.
- Silence progress-meter output when stdout is not a console.
- Unify warnings and errors so they go to stderr and are prefixed by "Warning:"
  or "Error:".
- Allow hex escapes that have bit 31 set.
- After cropping a sprite, compress and write the correct number of bytes.
- Add -X to print PCX sprite numbers in hex
- (devel) Many compilation issues with different compilers and architectures.
- Fix bug in UTF-8 generation for escapes \U800..\U7FFF.
- Fix U+E0xx characters being output quoted.
- (devel) By Tron: code cleanup, make C functions and variables static where
  possible, remove dead and debugging code.
- Require two slashes before dismissing anything as a C++ style comment.
- Fix reading of real sprites with Unixy absolute paths.
- Check for pixels with palette index of 255 when encoding.
- Add -q to silence warnings.
- Add -c to automatically crop extraenous transparent blue pixels.
- Add escapes for the new action 2 operations 0E, 0F, 11, 12, 13, 14, 15, 16
- Prevent quotes and escapes in sprite 0.
- Add escapes for long-format intro dates.

Changes in 0.9.10 (r819)
- (bugfix) Correct reading of \7c, \7C, \D| and \D&. (DaleStan)

Changes in 0.9.9 (r653)
- By Patchman:
  - Check for sprites that extend beyond the end of the PCX file.
  - Update versioning system to show last-comitted revision for non-release
    builds.
- By DaleStan:
  - (devel) Add release targets to the Makefiles
  - Check the palette of the PCX file before encoding from it.
    (Override a failed check with -f)
  - Introduce several backslash escape sequences; see the wiki for further
    information.
    (disable with -x or -xx)
  - Do not apply colormap to sprites that like character glyphs.
    (Apply colormap to all sprites with -M instead of -m)
  - Enable reading of GIMP palette files with -p.
  - (devel) Terse build mode; get the verbose mode with make V=1.
  - (devel) Touch version.h if and only if it changes.

Changes in 0.9.8 (r197)
- By DaleStan:
  - (devel) grfcodec (but not grfdiff or grfmerge) can now be built in
    MS Visual Studio 2003.
  - Slightly change the plaintext quoting algorithm.
  - (devel) Change several memcpy()s to pointer assignments.
  - (devel) Plug several memory leaks
  - (devel) Source code and Makefile changes to grfdiff and grfmerge build
    with both Makefiles, on both Cygwin and Linux.
  - Make code gcc 4 -Wall clean.
  - Use NFORenum's NFO reading engine:
    - Sprite numbers are now optional, and need not be correct if present.
    - Sprite lengths need not be correct. Set to 0 to supress the incorrect
      length warnings.
    - Real sprites are checked for format compliance.
  - Make 0-byte pseudo sprites an error.
  - (bugfix) Prevent binary includes from including directories on OSes that
    would othewise allow it.
  - (bugfix) It is now possible for a real sprite to start above the
    preceeding sprite (the infamous decreasing ypos bug)
  - grfcodec now exits with 0 on success.

Changes in 0.9.7
- support including binary files using "  4 ** data/dingdong.wav" syntax
- by Oskar: support // and # comments within pseudo-sprites
- fixed incorrect error message while encoding if .nfo file could not be read
  (now it reports "Cannot read %s" instead of Creating sprites/: File exists")
- made code -Wall clean
- remove myalloc stuff

Changes in 0.9.6
- by pasky:
  * made it possible to compile grfcodec on Linux
  * made compatible with gcc 3.x
  * when decoding, text strings are written in plain text (disable with -t)
  * supports plain text in .nfo files
  * fixed bug when first encoding a grf file wouldn't rename .new to .grf

Changes in 0.9.5
- self-extracting executables now work on Win2K etc., which doesn't pass
  the full path to the C library
- GRFMerge will just skip .grf files it can't find
- making sprites larger doesn't work if they're at the right edge; complains
  that ofsx too large; instead just use background colour
- GRFDiff doesn't like grfdiff -l 1254 ... (single value in -l), complains
  that there were no differences and makes no output
- allow GRFMerge to read GRFDiff's .exe files, both for -l and other options

Changes in 0.9.4
- fixed yet another crash while compressing
- sprites no longer need correct numbers in the .NFO file; just use "-1" for
  each number and it will be assumed to be correct

Changes in 0.9.3
- fixed crash due to buffer overrun while compressing
- slightly improved compression efficiency and speed
- GRFCodec now handles spaces in file/directory names properly
- GRFCodec now supports comments in .NFO files (from //, ; or # to end of line)
- speed up PCX processing

Changes in 0.9.2
- fixed crash if pcx file not found
- fixed crash if nfo file has no version info header

Changes in 0.9.1
- bug fix by Marcin: wrong order of test in encode function

Changes in release April 20 2002
- -p now actually uses the parameter you give it

Changes in release Dec 03 2001
- made -p option a valid option
- improve compression speed even more using assembly optimizations
- compression is now on by default, disable with -c-
- fixed bug where encoding would fail if the grf file didn't exist already
- fixed bug where decoding would fail creating the sprites directory
- corrected some mistakes in grf.txt (checksum, and initial offsets for tiles)

Changes in release March 26 2001
- fix bug that didn't flush last row if it only contained a single sprite
- fix bug with no height given (would make PCX have height -1)
- show percentage done while decoding
- fix compression algorithm
- improve compression performance
- try without compression if data is inconsistent after decompressing our

src/grfid.cpp
/*
 * This file is part of GRFCodec.
 * GRFCodec is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
 * GRFCodec is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include <string.h>

#include "md5.h"

#ifndef WIN32
#include <sys/mman.h>
#else
#include <stdlib.h>
#define PROT_READ 0
#define MAP_PRIVATE 0
/* Lets fake mmap for Windows, please! */
void *mmap (void *ptr, long size, long prot, long type, long handle, long arg) {
	char *mem = (char*)malloc(size + 1);
	mem[size] = 0;
	FILE *in = fdopen(handle, "rb");
	if (fread(mem, size, 1, in) != 1) {
		fclose(in);
		free(mem);
		return NULL;
	}
	fclose(in);
	return mem;
}
long munmap (void *ptr, long size) {
	free(ptr);
	return 0;
}
#endif

#include "version.h"

size_t _file_length;
uint8_t *_file_buffer;
uint8_t *_buffer;

inline void SkipBytes(size_t count)
{
	_buffer += count;
}

inline uint8_t ReadByte()
{
	return *(_buffer++);
}

inline uint16_t ReadWord()
{
	uint16_t v = ReadByte();
	return v | (ReadByte() << 8);
}

inline uint32_t ReadDWord()
{
	uint32_t v = ReadWord();
	return v | (ReadWord() << 16);
}

void SkipSpriteData(uint8_t type, uint16_t num)
{
	if (type & 2) {
		SkipBytes(num);
	} else {
		while (num > 0) {
			int8_t i = ReadByte();
			if (i >= 0) {
				int size = (i == 0) ? 0x80 : i;
				num -= size;
				SkipBytes(size);
			} else {
				i = -(i >> 3);
				num -= i;
				ReadByte();
			}
		}
	}
}

inline uint32_t Swap32(uint32_t x)
{
	return ((x >> 24) & 0xFF) | ((x >> 8) & 0xFF00) | ((x << 8) & 0xFF0000) | ((x << 24) & 0xFF000000);
}

static const char header[] = {
	'\x00', '\x00',                 // End-of-file marker for old OTTDp versions
	'G',    'R',    'F',    '\x82', // Container version 2
	'\x0D', '\x0A', '\x1A', '\x0A', // Detect garbled transmission
};

inline uint32_t ReadSize(int grfcontversion)
{
	return grfcontversion == 2 ? ReadDWord() : ReadWord();
}


const char *GetGrfID(const char *filename, uint32_t *grfid)
{
	*grfid = 0;

	FILE *f = fopen(filename, "rb");
	if (f == NULL) return "Unable to open file";

	/* Get the length of the file */
	fseek(f, 0, SEEK_END);
	_file_length = ftell(f);
	fseek(f, 0, SEEK_SET);

	/* Map the file into memory */
	_file_buffer = (uint8_t*)mmap(NULL, _file_length, PROT_READ, MAP_PRIVATE, fileno(f), 0);
	_buffer = _file_buffer;

	int grfcontversion = 1;

	if (_file_length > sizeof(header) && memcmp(_buffer, header, sizeof(header)) == 0) {
		grfcontversion = 2;
		_buffer += sizeof(header) + 4 + 1; // Header + offset till data + compression
	}

	/* Check the magic header, or what there is of one */
	if (ReadSize(grfcontversion) != 0x04 || ReadByte() != 0xFF) return "No magic header";

	/* Number of sprites. */
	ReadDWord();

	while (_buffer < _file_buffer + _file_length) {
		uint32_t num = ReadSize(grfcontversion);
		if (num == 0) break;

		uint8_t type = ReadByte();
		if (type == 0xFF) {
			/* Pseudo sprite */
			uint8_t action = ReadByte();
			if (action == 0x08) {
				/* GRF Version, do not care */
				ReadByte();
				/* GRF ID */
				*grfid = Swap32(ReadDWord());
				/* No more, we don't care */
				break;
			} else {
				SkipBytes(num - 1);
			}
		} else {
			SkipBytes(7);
			/* Skip sprite data */
			SkipSpriteData(type, num - 8);
		}
	}

	munmap(_file_buffer, _file_length);
	fclose(f);

	return (*grfid == 0) ? "File valid but no GrfID found" : NULL;
}

const char *GetMD5(const char *filename, md5_state_t *md5)
{
	FILE *f = fopen(filename, "rb");
	if (f == NULL) return "Unable to open file";

	/* Get the length of the file */
	fseek(f, 0, SEEK_END);
	_file_length = ftell(f);
	fseek(f, 0, SEEK_SET);

	/* Map the file into memory */
	_file_buffer = (uint8_t*)mmap(NULL, _file_length, PROT_READ, MAP_PRIVATE, fileno(f), 0);
	_buffer = _file_buffer;

	size_t read_length = _file_length;
	if (_file_length > sizeof(header) && memcmp(_buffer, header, sizeof(header)) == 0) {
		_buffer += sizeof(header);
		/* Reduce the length to contain only the data, but not the sprites. */
		read_length = sizeof(header) + 4 + ReadDWord();
	}

	md5_append(md5, _file_buffer, read_length);

	munmap(_file_buffer, _file_length);
	fclose(f);

	return NULL;
}

int main(int argc, char **argv)
{
	if (argc < 2 || strcmp(argv[1], "-h") == 0 || (strcmp(argv[1], "-m") == 0 && argc < 3)) {
		printf(
			"GRFID " VERSION " - Copyright (C) 2009 by Peter Nelson\n"
			"\n"
			"Usage:\n"
			"    GRFID <NewGRF-File>\n"
			"        Get the GRF ID from the NewGRF file\n"
			"    GRFID -m <NewGRF-File>\n"
			"        Get the MD5 checksum of the NewGRF file\n"
			"    GRFID -v\n"
			"        Get the version of GRFID\n"
			"\n"
			"You may copy and redistribute it under the terms of the GNU General Public\n"
			"License, as stated in the file 'COPYING'.\n"
			"Uses MD5 implementation of L. Peter Deutsch.\n");
		return 1;
	}
	if (strcmp(argv[1], "-v") == 0) {
		printf("GRFID " VERSION " - Copyright (C) 2009 by Peter Nelson\n");
		return 0;
	}

	const char *err;
	if (strcmp(argv[1], "-m") == 0) {
		md5_state_t md5;
		md5_init(&md5);
		err = GetMD5(argv[2], &md5);
		if (err == NULL) {
			md5_byte_t digest[16];
			md5_finish(&md5, digest);
			for (size_t i = 0; i < sizeof(digest); i++) {
				printf("%02x", digest[i]);
			}
			printf("\n");
			return 0;
		}
	} else {
		uint32_t grfid;
		err = GetGrfID(argv[1], &grfid);

		if (err == NULL) {
			printf("%08x\n", grfid);
			return 0;
		}
	}

	printf("Unable to get requested information: %s\n", err);
	return 1;
}

